### PR TITLE
Fix Invalid foreign key when applying schema snapshot

### LINKED ISSUE #9723: Schema snapshot is not taking foreign keys into account
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

When applying [this schema](https://gist.github.com/NecoHorne/55f57824e68083c9baaabfce1c2eb856) it fails with the following error:
```sh
08:25:32 🚨 Invalid foreign key in field "group".
Error: Invalid foreign key in field "group".
    at foreignKeyViolation (/Users/josevarela/workspace/directus/directus/api/src/exceptions/database/dialects/postgres.ts:116:9)
    at extractError (/Users/josevarela/workspace/directus/directus/api/src/exceptions/database/dialects/postgres.ts:28:11)
    at translateDatabaseError (/Users/josevarela/workspace/directus/directus/api/src/exceptions/database/translate.ts:28:27)
    at /Users/josevarela/workspace/directus/directus/api/src/services/items.ts:141:39
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

And the query associated:
```
  knex:query insert into "directus_collections" ("accountability", "archive_app_filter", "archive_field", "archive_value", "collapse", "collection", "color", "display_template", "group", "hidden", "icon", "item_duplication_fields", "note", "singleton", "sort", "sort_field", "translations", "unarchive_value") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18) returning "collection" trx242 +2ms
  knex:bindings [
  knex:bindings   'all',
  knex:bindings   true,
  knex:bindings   'status',
  knex:bindings   'archived',
  knex:bindings   'open',
  knex:bindings   'docs',
  knex:bindings   null,
  knex:bindings   '{{title}} — {{status}}',
  knex:bindings   'sections',
  knex:bindings   false,
  knex:bindings   'school',
  knex:bindings   null,
  knex:bindings   'Documentation — Articles',
  knex:bindings   false,
  knex:bindings   1,
  knex:bindings   'sort',
  knex:bindings   null,
  knex:bindings   'draft'
  knex:bindings ] trx242 +2ms
  ```

### To Reproduce

- Download the above schema
- Apply it `npx directus schema apply ./snapshot.yaml`

### Errors Shown

_No response_

### What version of Directus are you using?

9.0.0 (main)

### What version of Node.js are you using?

16.13.0

### What database are you using?

Postgres 13

### What browser are you using?

Edge

### What operating system are you using?

macOS

### How are you deploying Directus?

Locally

### PR DESCRIPTION
**Pull request by PeeraJ**
Fix #9723 
it's my first time contributing back to Directus community please let me know if something needs to be changed. 
image

as of before if you have data models as shown when you try to apply schema it will throw an error like this
image

after fix
image

### CODE REVIEW

**Comment by eyecatchup** [comment]
This is a real deal breaker. @rijkvanzanten Could you gimme a rough estimate when you will have time to review this PR? Thanks.

File: api/src/utils/apply-snapshot.ts [hunk 1]
```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
...
-		for (const { collection, diff } of snapshotDiff.collections) {
-			if (diff?.[0].kind === 'D') {
-				try {
-					await collectionsService.deleteOne(collection);
-				} catch (err) {
-					logger.error(`Failed to delete collection "${collection}"`);
-					throw err;
+		const createCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
+				for (const item of collections) {
```
Comment (azrikahar) [inline api/src/utils/apply-snapshot.ts:58]:
```suggestion
				for (const { collection, diff } of collections) {
```

More of a nitpick but let's keep the original destructuring to keep things similar as before 👍 (unless of course there's a naming collision that I'm missing here)

File: api/src/utils/apply-snapshot.ts [hunk 2]
```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
...
-						...diff[0].rhs,
-						fields,
-					});
-				} catch (err: any) {
-					logger.error(`Failed to create collection "${collection}"`);
-					throw err;
+		};
+		const deleteCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
+				for (const item of collections) {
```
Comment (azrikahar) [inline api/src/utils/apply-snapshot.ts:82]:
```suggestion
				for (const { collection, diff } of collections) {
```

**Review by azrikahar — CHANGES_REQUESTED** [review]

**Comment by eyecatchup** [comment]
I've added a [PR](https://github.com/PeeraJ/directus/pull/1/commits) to the patch branch to fix the deconstructing.

**Comment by eyecatchup** [comment]
I see there're some unanswered questions from your initial review. @azrikahar could you please clarify if you're waiting for further specific answers from @PeeraJ or if you just not had the time to review the updates to this PR?

**Comment by azrikahar** [comment]
> could you please clarify if you're waiting for further specific answers from @PeeraJ or if you just not had the time to review the updates to this PR?

Honestly both! If PeeraJ is still busy and time permits on my end, I'll try to add the remaining tweaks and merge conflict before the next planned release. Though ultimately it can be rather hard to give any certainty with other things taking priority at the moment. Hope that's understandable. 👍

**Comment by eyecatchup** [comment]
Sure thing, understandable. Thanks for the quick reply! I've [moved the type declaration to the top](https://github.com/PeeraJ/directus/pull/2), hopefully @PeeraJ can merge this soon. Are there any remaining "tweaks" you would need? Otherwise I could take care of resolving the merge conflicts after my PR was merged to get this fixed asap, without blocking your time.

**Comment by PeeraJ** [comment]
> Sure thing, understandable. Thanks for the quick reply! I've [moved the type declaration to the top](https://github.com/PeeraJ/directus/pull/2), hopefully @PeeraJ can merge this soon. Are there any remaining "tweaks" you would need? Otherwise I could take care of resolving the merge conflicts after my PR was merged to get this fixed asap, without blocking your time.

I'm so sorry to leave this unattended, I just came back from a long vacation and am unable to fix those problems right away. Thanks a lot for the pr though.🙏

File: api/src/utils/apply-snapshot.ts [hunk 3]
```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
...
 
-		for (const { collection, diff } of snapshotDiff.collections) {
-			if (diff?.[0].kind === 'D') {
-				try {
-					await collectionsService.deleteOne(collection);
-				} catch (err) {
-					logger.error(`Failed to delete collection "${collection}"`);
-					throw err;
+		const createCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
```
[the remark this answers is not in the document]
Comment (PeeraJ) [reply api/src/utils/apply-snapshot.ts]: removed.

**Comment by eyecatchup** [comment]
Great. Thanks @PeeraJ for merging my change and removing the unnecessary empty check. That being done, it only needs to resolve the merge conflicts. Since I can not do it on this PR, I needed to create a duplicate. This PR https://github.com/directus/directus/pull/13096 includes all the changes from PeeraJ and me and resolves the merge conflict. @azrikahar anything else I can assist with to get this merged?

**Comment by azrikahar** [comment]
@PeeraJ no worries! Thanks for updating promptly. I'll take it from here if that's okay with you 👍

> anything else I can assist with to get this merged?

@eyecatchup all good. I'd like to lean on to the safer side and do some tests on my end just to prevent any oversight here. I've resolved the merge conflict for now, and I'll add on any changes if needed before wrapping this up. Thanks again for helping out.

**Review by azrikahar — APPROVED** [review]

**Pull request merged**